### ISSUE TITLE
Slow toggling of column visibility with many columns and horizontal virtual dom

### ISSUE THREAD

**Issue by blankbadge** [olifolkerd/tabulator#3741]
Hi Oli,

This is a slightly modified re-report of this issue:

https://github.com/olifolkerd/tabulator/issues/3728

I am filing again as a separate issue because you closed the original issue and suggested to use horizontal virtual dom to improve handling of large number of columns.  But even with that enabled I still see the issue, so it doesn't seem to be caused by lack of virtual dom.

**Describe the bug**
I am encountering an issue with tables containing larger numbers of columns. The performance when toggling the visibility of columns (col.show(), col.hide(), or col.toggle()) degrades with larger number of columns. I was expecting a fixed runtime to show/hide each column, for example 1ms per column. But what I find is that the more columns, the more time each column takes to toggle (it might instead take 10ms per column with more total columns).  Even with horizontal virtual dom enabled, which you suggested in the original issue, this problem persists.  In fact, enabling horizontal vdom actually increased the time it takes to toggle visibility on each column until you get to around 80-100 columns.

Here are some metrics from testing both basic and virtual horizontal dom with differing number of columns.  Times shown are the times to toggle visibility per column, not the total time.  The pattern shows that at lower number of columns, the basic dom is faster than the virtual dom for toggling visibility.  I don't know what size buffer you use for horizontal vdom, but I'm assuming there can be some expected overhead here in managing the vdom and so we see a performance penalty with a small number of columns where vdom isn't really needed.  That's okay.  

But at larger columns where vdom should show its benefit, the typical time to toggle column visibility is similar with both basic and virtual dom, and that time increases per-column as more columns are added, making it impractical to do this on a table with a very large number of columns.

Columns | Basic | Virtual
-- | -- | --
5 | 2ms | 4ms
10 | 4ms | 8ms
20 | 7ms | 15ms
40 | 22ms | 25ms
60 | 35ms | 35ms
80 | 50ms | 50ms
100 | 75ms | 73ms

**Tabulator Info**
Tabulator v5.2.4

**Working Example**
This fiddle shows the issue. You can change the numCols variable at the beginning of the script to test with different number of columns.  You can also change the renderHorizontal variable between "basic" and "virtual" to test with the default behavior or with horizontal virtual dom.

https://jsfiddle.net/blankbadge/w60ychzo/

**To Reproduce**
Run the testcase with different numbers of columns using the numCols variable.  As you go up from 5 columns to 100, you'll see the time it takes to toggle visibility of each column increases.  This increase is seen whether or not horizontal virtual dom is enabled (you can toggle the renderHorizontal variable between "basic" and "virtual" to test both ways).

**Expected behavior**
Expectation is that with horizontal virtual dom enabled, increasing the number of columns does not increase the time to toggle each individual column.  I would expect for example that toggling 200 columns takes about twice as long as toggling 100 columns, but not that it would scale non-linearly and take 3x or 4x longer.  I understand that this may only be true for those columns which are virtualized and not the ones currently displayed in the real dom.  I also do not expect this improvement if horizontal virtual dom is disabled, as you explained in the other issue.  No virtual dom with many columns can have performance issues, that is okay.

**Desktop (please complete the following information):**
Windows 10 with Chrome 100

**Comment by olifolkerd** [comment]
Hey @blankbadge 

Thanks for the info,

It definitely does need to be the horizontal virtual renderer for efficiency, that is the whole point of it. the basic renderer will result in all the columns being redrawn when changes are made which is why it is so in efficient.

The horizontal renderer, is still in the experimental stages and has only really been tested on static table data, which it renders much faster than the basic renderer (case in point is the Beekeeper Studios app that uses that renderer extensively for its performance improvements), but currently doesn't handle any changes in column visibility, which is likely why you are seeing the numbers you are seeing.

What i would say is if you are planning on doing complex multi column manipulation, then you should block the redraw before you start manipulating the DOM. that way only one redraw will occur when all your updates are completed, which will run much faster:

```js
table.blockRedraw();
  for (const col of table.getColumns()) {
    if (!firstCol) {
      col.toggle();
    }
    firstCol = false;
  }
table.restoreRedraw();
```

But if you are only doing all of that to get an average for bench marking then that wont offer you any real life benefits.

I will have a tinker with the renderer and see what can be improved for the next release.

Cheers

Oli :)

**Comment by blankbadge** [comment]
Thanks for the advice, Oli!  Really appreciate it.  I have played around with blockRedraw and restoreRedraw in the past for this situation in my actual application (toggling visibility of many columns) and didn't see any notable improvement.  I just tested on the jsfiddle above.  I used 120 columns to hopefully really show the benefit of both the virtual dom and blocking redraw.  I did a simple test where I ran each of these configurations seven times, dropped the fastest/slowest times and averaged the remaining five.  The results are:

| Test | Time per Column |
| --- | --- |
| Basic, No Blocking | 83ms |
| Basic, Blocking | 76ms |
| Horiz Vdom, No Blocking | 60ms |
| Horiz Vdom, Blocking | 68ms |

The results show that enabling the horizontal vdom improves performance, from 83ms to 60ms without blocking redraw between each column hide.  But surprisingly, blocking redraw is actually slowing down the horizontal vdom, from 60ms to 68ms.  

Just sharing the data in case that gives you any helpful clues.  But I think these differences of time from 60ms-83ms aren't showing the real issue.  I suspect a vast majority of this runtime is not from table rendering.  Because:

For horizontal vdom, I have 120 columns and display width is set to only show about five at a time.  So assuming the vdom renders double that (10 columns), that means there were still as many as 110 columns that vdom would not have rendered.  So I would expect more than a 10-20% performance improvement would be seen.

Similarly for blocking redraw: Hiding 120 columns one by one would trigger 120 redraws without blocking, but only one redraw with blocking, right?  So if the performance of the redraws was the bottleneck, again we should see a very dramatic improvement, but it's only about 10% difference (and blocking actually slows things down with the horiz vdom enabled).

Instead, almost all of this runtime must be coming from somewhere else.  Which is back to the original issue: performance per column degrades linearly with the number of columns in the table.  If I drop the number of columns down to 10, it doesn't matter if I use virtual or basic rendering, or whether I block redraw on every column hide -- they all perform at around 2-6ms per column instead of the 60-80ms per column with 120 columns.

By the way, I'm not benchmarking this performance just to be academic.  This is one of the two main performance bottlenecks I've identified in my own application.  In my usage case, I'm creating tables that could be 100+ columns wide, but in practice I'm almost always limiting it to < 50 columns because of performance with showing/hiding columns.  It's a common action in my application, because the user typically would focus on the data from only a few of those columns at once, and want to hide the other columns.  So this "toggle off visibility for 40 columns" type of task is common.

If there was a way to do a batch update of multiple column visibility at once, or otherwise eliminate the slowdown with increasing column count, that would be great!  Thanks for considering this request :)

thanks!
blankbadge

**Comment by olifolkerd** [comment]
Hey @blankbadge 

Thanks for the detailsed feedback, that was very helpful,

Yes it would appear that when i was implementing the `blockRedraw` functionality that i only did the row side of it not the column side of it.

That should actually be a really quick fix to implement. I'm planning on releasing v5.3 in a week or so,  i will ensure that it includes the ability to block column updates too.

Cheers

Oli :)

**Comment by olifolkerd** [comment]
Hey @blankbadge 

Great news, i have fixed the issue with the redraw scaling with column size so columns can now individually be toggled in about 20ms. I have also extended the `blockRedraw` function to the column manager, with that enabled you are loing at 0.08ms per column to toggle visibility.

That should hopefully speed things up for you.

This will be included in the 5.3 release in a week or so.

Cheers

Oli :)

**Comment by blankbadge** [comment]
Great, appreciate the quick help, Oli!  Right now I’m not using the horizontal vdom because of the restriction in handling frozen columns (which are used in all my tables) — you had mentioned on another issue that frozen column support with horiz vdom was already on your enhancement list for the coming months. So looking forward to that. But from your description, sounds like this fix will give a big performance boost even without the horiz vdom enabled if I use blockRedraw, so I will definitely try it out when 5.3 is released. 

Thanks!
blankbadge
